pkg/doc/jsonld/document_loader.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package jsonld
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/piprate/json-gold/ld"
+)
+
+// CachingDocumentLoader is similar to json-gold's CachingDocumentLoader but uses cache as *sync.Map instead.
+type CachingDocumentLoader struct {
+	nextLoader ld.DocumentLoader
+	cache      map[string]interface{}
+	rwMutex    sync.RWMutex
+}
+
+// NewCachingDocLoader creates a new instance of CachingDocumentLoader.
+func NewCachingDocLoader(nextLoader ld.DocumentLoader) *CachingDocumentLoader {
+	cdl := &CachingDocumentLoader{
+		nextLoader: nextLoader,
+		cache:      map[string]interface{}{},
+	}
+
+	return cdl
+}
+
+// LoadDocument returns a RemoteDocument containing the contents of the JSON resource from the given URL (u).
+func (cdl *CachingDocumentLoader) LoadDocument(u string) (*ld.RemoteDocument, error) {
+	cdl.rwMutex.RLock()
+	if doc, cached := cdl.cache[u]; cached {
+		defer cdl.rwMutex.RUnlock()
+
+		if cachedDoc, ok := doc.(*ld.RemoteDocument); ok {
+			return cachedDoc, nil
+		}
+
+		return nil, errors.New("invalid document entry")
+	}
+
+	cdl.rwMutex.RUnlock()
+
+	docFromLoader, err := cdl.nextLoader.LoadDocument(u)
```

Comment:
Looks like `nextLoader` must also be thread-safe.

This will be an issue when people want to add extra contexts in downstream projects and they understandably forget about this bug, such as when they use the [`DocumentVerifier`](https://github.com/hyperledger/aries-framework-go/blob/fea319ee4b47b0a81ceea9f2e6d96e3902c5d1f1/pkg/doc/signature/verifier/verifier.go#L70-L79) directly (like zcaps do).

The better solution is to have the user prepare the contexts themselves in an `ld.DocumentLoader` and be done with this. No need for locking or any of this extra complexity.